### PR TITLE
Prune utils modules

### PR DESCRIPTION
**Pull request by alex-mckenna**
Pruning `Clash.Utils` and `Clash.Core.Utils` as they both contain definitions which are redundant or can be moved to a more focused module. This is just a quick-fire improvement, intended to make it easier to improve the `Clash.Core` namespace later.

### CODE REVIEW

File: clash-lib/src/Clash/Core/Subst.hs [hunk 1]
```diff
@@ -239,7 +239,7 @@ zipTvSubst
   -> Subst
 zipTvSubst tvs tys
   | debugIsOn
-  , neLength tvs tys
+  , length tvs /= length tys
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Core/Subst.hs]: Depending on what magic GHC applies, the new version does a lot more work. (Might be a microoptimization though.)

File: clash-lib/src/Clash/Core/Subst.hs [hunk 2]
```diff
@@ -610,7 +610,7 @@ substTyWith
   -> Type
   -> Type
 substTyWith tvs tys =
-  ASSERT( tvs `equalLength` tys )
+  ASSERT( length tvs == length tys )
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Core/Subst.hs]: Same

File: clash-lib/src/Clash/Util.hs [hunk 3]
```diff
@@ -295,24 +254,6 @@ indexNote
...
 
--- | Safe version of 'head'
-headMaybe :: [a] -> Maybe a
-headMaybe (a:_) = Just a
-headMaybe _ = Nothing
-
--- | Safe version of 'tail'
-tailMaybe :: [a] -> Maybe [a]
-tailMaybe (_:as) = Just as
-tailMaybe _ = Nothing
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Util.hs]: IMO keep these around, they're useful no?

File: clash-lib/src/Clash/Util.hs [hunk 4]
```diff
@@ -447,12 +366,6 @@ allM p (x:xs) = do
   else
     return False
 
-traceWith :: (a -> String) -> a -> a
-traceWith f a = trace (f a) a
-
-traceShowWith :: Show b => (a -> b) -> a -> a
-traceShowWith f a = trace (show (f a)) a
-
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Util.hs]: These are super useful when debugging!

File: clash-lib/src/Clash/Util.hs [hunk 5]
```diff
@@ -295,24 +254,6 @@ indexNote
...
-headMaybe (a:_) = Just a
-headMaybe _ = Nothing
-
--- | Safe version of 'tail'
-tailMaybe :: [a] -> Maybe [a]
-tailMaybe (_:as) = Just as
-tailMaybe _ = Nothing
-
--- | Split the second list at the length of the first list
-splitAtList :: [b] -> [a] -> ([a], [a])
```
Comment (christiaanb) [inline clash-lib/src/Clash/Util.hs]: Why remove this? `splitAtList xs ys` is faster than `splitAt (lenght xs) ys` when `xs` is bigger than `ys`, as you don't have to fully traverse `xs`

File: clash-lib/src/Clash/Util.hs [hunk 6]
```diff
@@ -348,20 +289,6 @@ clogBase x y | x > 1 && y > 0 =
                 else Just (I# z1)
 clogBase _ _ = Nothing
 
--- | Determine whether two lists are of equal length
-equalLength
```
Comment (christiaanb) [inline clash-lib/src/Clash/Util.hs]: Same here, why remove this when `equalLength xs ys` finishes earlier than `lenght xs == length ys`

File: clash-lib/src/Clash/Util.hs [hunk 7]
```diff
@@ -348,20 +289,6 @@ clogBase x y | x > 1 && y > 0 =
...
-  :: [a] -> [b] -> Bool
-equalLength [] [] = True
-equalLength (_:as) (_:bs) = equalLength as bs
-equalLength _ _ = False
-
--- | Determine whether two lists are not of equal length
-neLength
```
Comment (christiaanb) [inline clash-lib/src/Clash/Util.hs]: Same question as `splitAtList` and `equalLength`

File: clash-lib/src/Clash/Util.hs (hunk 3 above)
Comment (alex-mckenna) [reply clash-lib/src/Clash/Util.hs]: They were only used in one place and both were used (equivalent to uncons).
I also don't think `fmap fst . uncons` etc. is any more work to write :man_shrugging:

Comment (martijnbastiaan) [reply clash-lib/src/Clash/Util.hs]: I wasn't aware of `uncons` (and it certainly a  better fit for where it's used). Having said that, `headMaybe` is definitely easier to write, but more importantly read, than `fmap fst . uncons`.

File: clash-lib/src/Clash/Util.hs (hunk 4 above)
Comment (alex-mckenna) [reply clash-lib/src/Clash/Util.hs]: Fair point, what about a debugging module that exports these (e.g. `Clash.Internal.Debug`)?

File: clash-lib/src/Clash/Util.hs (hunk 3 above)
Comment (alex-mckenna) [reply clash-lib/src/Clash/Util.hs]: True, but there's also the extra where do I import that from / where is that imported from which is avoided with `uncons`. Given it was only used in one place I don't think this is much of a loss anyway

**Comment by alex-mckenna** [comment]
With the removed functions that @christiaanb mentioned, I think removing them from utils is still a good idea as

  * each was only used one
  * `equalLength` is only used in an `ASSERT`, I don't think the increase in speed is important given it only occurs in debug builds anyway
 
With `splitAtList`, it might be a good idea to add it back, as the function that uses it is used in `coreView1`. Although I think it should be added back locally to the function that uses it's `where` block if this is what we want to do

**Comment by alex-mckenna** [comment]
Discussion offline, the list functions should be kept for now, but will be moved to a more appropriate module (i.e. `Data.List.Extra`, but feel free to bikeshed :wink:).

File: clash-lib/src/Clash/Util.hs (hunk 4 above)
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Util.hs]: The nice thing about these is that they're usually imported by default as many modules import `Clash.Util` unqualified (eww..).

File: clash-lib/src/Clash/Core/Term.hs [hunk 8]
```diff
@@ -126,6 +139,57 @@ data Pat
...
+patIds _                    = ([],[])
+
+patVars :: Pat -> [Var a]
+patVars (DataPat _ tvs ids) = coerce tvs ++ coerce ids
+patVars _ = []
+
+-- | Abstract a term over a list of term and type variables
+mkAbstraction :: Term
+              -> [Either Id TyVar]
+              -> Term
```
Comment (martijnbastiaan) [inline clash-lib/src/Clash/Core/Term.hs]: If you feel like it: reformat these functions to adhere to the styleguide (i.e., type signature on a single line OR the `::` on the next line.

Comment (alex-mckenna) [reply clash-lib/src/Clash/Core/Term.hs]: I always feel like making the code prettier, I just didn't want to make all the diffs confusing
If this gets approved, how about I squash the two pruning commits and then do the stylistic changes on the squashed commit?

Comment (martijnbastiaan) [reply clash-lib/src/Clash/Core/Term.hs]: Works for me, no need to squash IMO.

Comment (alex-mckenna) [reply clash-lib/src/Clash/Core/Term.hs]: Fair, in that case I'll leave a style fix commit at the end of the PR so the history doesn't become more tedious to read

**Pull request merged**